### ISSUE TITLE
MultiPolygon Intersection Time

### ISSUE THREAD

**Issue by BebeSparkelSparkel** [Toblerity/Shapely#483]
I have conducted a simple time test of your MultiPolygon intersections. It seems to work quickly if there is no intersection. However, it is much slower (0.0017s vs 0.00008s) than a simple comparison algorithm I created if there is an intersection.

I'd be willing to try and speed this up but have not found where you implement the intersection algorithm. Is it a part of the geos library?

Comparison timings boxed in red.
![screen shot 2017-04-18 at 7 42 20 am](https://cloud.githubusercontent.com/assets/10747532/25129216/12ceafac-240b-11e7-9129-3b76ea493842.png)
```
label                         min          max         mean       total    run count
--------------------  -----------  -----------  -----------  ----------  -----------
no_intersect_any      0.00237799   0.00502706   0.00273036   1.36518             500
intersect_multi       0.00152588   0.00437808   0.00172278   0.861392            500
no_intersect_multi_2  0.00115705   0.00272393   0.00131113   0.655567            500
intersect_any         1.69277e-05  0.00100398   8.48861e-05  0.042443            500
no_intersect_multi    1.4782e-05   8.60691e-05  1.8177e-05   0.00908852          500
```

Patterns I used  to check the intersections and non-intersections with.
![multipolygon_intersection](https://cloud.githubusercontent.com/assets/10747532/25128832/4e0e3dfa-2409-11e7-8ebe-53bc8feaffb6.png)

**Comment by snorfalorpagus** [comment]
The intersection in Shapely is a very thin wrapper around the intersection method in GEOS.

It's a little hard to follow if you're not familiar with the code base. The relevant lines are here:
https://github.com/Toblerity/Shapely/blob/b4c11357825f05acade1e9de3f4f1f858dd700f9/shapely/geometry/base.py#L618
https://github.com/Toblerity/Shapely/blob/7b09eb777f688d9c404b0f20b80864c8e51c446e/shapely/impl.py#L90
https://github.com/Toblerity/Shapely/blob/1dad4d7ea80d2b0d55c7965a43648ef0e8060d1d/shapely/geos.py#L633
https://github.com/Toblerity/Shapely/blob/master/shapely/ctypes_declarations.py#L180

It really just passes the two pointers to GEOSIntersects, then wraps the result in a Shapely geometry object.

https://github.com/OSGeo/geos/blob/svn-trunk/capi/geos_ts_c.cpp#L488

The time spent actually in Shapely should be the same regardless of the result of the operation so if you want to speed things up you're going to have to do it in GEOS.
